File: app/src/main/java/com/vibhorsrv/cameraids/model/CameraModel.java
package com.vibhorsrv.cameraids.model;

import androidx.annotation.NonNull;

import java.util.Arrays;
import java.util.Locale;

public class CameraModel {
    private final int id;
    private CameraType cameraType;
    private float zoomFactor;
    private Camera2ApiProperties camera2ApiProperties;
    private DerivedProperties derivedProperties;

    public CameraModel(int id) {
        this.id = id;
    }

    public float getZoomFactor() {
        return zoomFactor;
    }

    public void setZoomFactor(float zoomFactor) {
        this.zoomFactor = zoomFactor;
    }

    public CameraType getCameraType() {
        return cameraType;
    }

    public void setCameraType(CameraType cameraType) {
        if (this.cameraType == null) {
            this.cameraType = cameraType;
        }
    }

    public int getId() {
        return id;
    }

    public Camera2ApiProperties getCamera2ApiProperties() {
        return camera2ApiProperties;
    }

    public void setCamera2ApiProperties(Camera2ApiProperties camera2ApiProperties) {
        this.camera2ApiProperties = camera2ApiProperties;
    }

    public DerivedProperties getDerivedProperties() {
        return derivedProperties;
    }

    public void setDerivedProperties(DerivedProperties derivedProperties) {
        this.derivedProperties = derivedProperties;
    }

    public boolean isTypeSet() {
        return cameraType != null;
    }

    @Override
    @NonNull
    public String toString() {
        return "\n" +
                "CameraID = [" + id + ']' + (!cameraType.equals(CameraType.LOGICAL) ? "  \u2605" : (!camera2ApiProperties.getPhysicalIds().isEmpty() ? " = " + camera2ApiProperties.getPhysicalIds().toString().replace(", ", "+") : "")) +
                "\n" +
                "Facing = " + derivedProperties.getFacing() +
                "\n" +
                (!cameraType.equals(CameraType.LOGICAL) ?
                        "Zoom = " + String.format(Locale.ROOT, "%.2fx", zoomFactor).replace(".00", "") +
                                "\n" : "") +
                "Type = " + cameraType +
                "\n" +
                "FocalLength = " + String.format(Locale.ROOT, "%.2fmm", camera2ApiProperties.getFocalLength()) +
                "\n" +
                "35mm eqv FocalLength = " + String.format(Locale.ROOT, "%.2fmm", derivedProperties.getMm35FocalLength()) +
                "\n" +
                "Aperture = " + camera2ApiProperties.getAperture() +
                "\n" +
                "SensorSize = " + camera2ApiProperties.getSensorSize().toString() +
                "\n" +
                "PixelArray = " + camera2ApiProperties.getPixelArraySize().toString() +
                "\n" +
                "PixelSize = " + String.format(Locale.ROOT, "%.2f", derivedProperties.getPixelSize()) + "µm" +
                "\n" +
                "AngleOfView(Diagonal) = " + Math.round(derivedProperties.getAngleOfView()) + "\u00b0" +
                "\n" +
                "AEModes = " + Arrays.toString(camera2ApiProperties.getAeModes()) +
                "\n" +
                "FlashSupported = " + camera2ApiProperties.isFlashSupported() +
                "\n" +
                "RAW_SENSOR sizes = " + Arrays.deepToString(camera2ApiProperties.getRawSensorSizes()) +
                "\n" +
                "SupportedHardwareLevel = " + camera2ApiProperties.getSupportedHardwareLevelString() +
                "\n";
    }
}

File: app/src/main/java/com/vibhorsrv/cameraids/model/CameraModel.kt
package com.vibhorsrv.cameraids.model

import java.util.Locale
import kotlin.math.roundToInt

class CameraModel(val id: Int) {
    var cameraType: CameraType? = null
    var zoomFactor = 0f
    lateinit var camera2ApiProperties: Camera2ApiProperties
    lateinit var derivedProperties: DerivedProperties

    val isTypeSet: Boolean
        get() = cameraType != null

    override fun toString(): String {
        return mutableMapOf<String, String>().apply {
            this["CameraID"] = "[$id]"
            if (cameraType != CameraType.LOGICAL) {
                this["CameraID"] += "  \u2605"
            } else if (camera2ApiProperties.physicalIds.isNotEmpty()) {
                this["CameraID"] +=
                    " = ${camera2ApiProperties.physicalIds.toString().replace(", ", "+")}"
            }

            this["Facing"] = "${derivedProperties.facing}"

            if (cameraType != CameraType.LOGICAL) {
                this["Zoom"] = String.format(Locale.ROOT, "%.2fx", zoomFactor)
                    .replace(".00", "")
            }

            this["Type"] = "$cameraType"
            this["FocalLength"] = String.format(
                Locale.ROOT, "%.2fmm", camera2ApiProperties.focalLength)
            this["35mm eqv FocalLength"] = String.format(
                Locale.ROOT, "%.2fmm", derivedProperties.mm35FocalLength)
            this["Aperture"] = "${camera2ApiProperties.aperture}"
            this["SensorSize"] = "${camera2ApiProperties.sensorSize}"
            this["PixelArray"] = "${camera2ApiProperties.pixelArraySize}"
            this["PixelSize"] = String.format(Locale.ROOT, "%.2fµm", derivedProperties.pixelSize)
            this["AngleOfView(Diagonal)"] = "${derivedProperties.angleOfView.roundToInt()}°"
            this["AEModes"] = camera2ApiProperties.aeModes.contentToString()
            this["FlashSupported"] = "${camera2ApiProperties.isFlashSupported}"
            this["RAW_SENSOR sizes"] = camera2ApiProperties.rawSensorSizes.contentDeepToString()
            this["SupportedHardwareLevel"] = camera2ApiProperties.supportedHardwareLevelString
        }.flatMap { (k, v) -> listOf("$k = $v") }.joinToString("\n")
    }
}
